packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,153 @@
+      return [serviceDefinitions, true];
+    });
+    update.mockImplementationOnce(async (config: GatewayConfig) => {
+      const services = serviceDefinitions.filter(s => s.name !== 'books');
+      // overwrite books service with a similar 'book' service
+      return [
+        [
+          ...services,
+          {
+            name: 'book',
+            typeDefs: books.typeDefs,
+            url: 'http://localhost:32542',
+          },
+        ],
+        true,
+      ];
+    });
+
+    const experimental_didUpdateComposition = jest.fn();
+
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: update,
+      experimental_pollInterval: 10,
+      experimental_didUpdateComposition,
+    });
+
+    const { interval } = await gateway.load();
+    await new Promise(resolve => setTimeout(resolve, 20));
+
+    clearInterval(interval);
+    const {
+      calls: [firstCall, secondCall],
+    } = experimental_didUpdateComposition.mock;
+
+    expect(experimental_didUpdateComposition).toHaveBeenCalledTimes(2);
+
+    // first call's `current` arg should have a schema
+    expect(firstCall[0].schema).toBeDefined();
+    expect(firstCall[1]).toBeUndefined();
+
+    // second call should have a previous schema
+    expect(secondCall[0].schema).toBeDefined();
+    expect(secondCall[1].schema).toBeDefined();
+    done();
+  });
+
+  it('uses default service definition updater', async () => {
+    const gateway = new ApolloGateway({
+      localServiceList: serviceDefinitions,
+    });
+
+    const { schema } = await gateway.load();
+
+    // spying on gateway.loadServiceDefinitions wasn't working, so this also
+    // should test functionality. If there's no overwriting service definition
+    // updater, it has to use the default. If there's a valid schema, then
+    // the loader had to have been called.
+    expect(schema.getType('User')).toBeDefined();
+  });
+
+  it('warns when polling on the default fetcher', async () => {
+    const consoleSpy = jest.spyOn(console, 'warn');
+    const gateway = new ApolloGateway({
+      serviceList: serviceDefinitions,
+      experimental_pollInterval: 10,
+    });
+    expect(consoleSpy).toHaveBeenCalledTimes(1);
+    expect(consoleSpy.mock.calls[0]).toMatchInlineSnapshot(`
+      Array [
+        "Polling running services is dangerous and not recommended in production. Polling should only be used against a registry. If you are polling running services, use with caution.",
+      ]
+    `);
+    consoleSpy.mockRestore();
+  });
+
+  it('warns when polling using a custom serviceList fetcher', async () => {
+    const consoleSpy = jest.spyOn(console, 'warn');
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: jest.fn(),
+      experimental_pollInterval: 10,
+    });
+    expect(consoleSpy).toHaveBeenCalledTimes(1);
+    expect(consoleSpy.mock.calls[0]).toMatchInlineSnapshot(`
```

Comment:
Also worried if we want to be a bit less restrictive about snapshot words, since this means changing wording means updating lots of tests.